wallet/src/Pos/Wallet/Web/Mode.hs

```diff
@@ -87,11 +97,46 @@ import           Pos.WorkMode                   (RealModeContext (..))
 data WalletWebModeContext = WalletWebModeContext
     { wwmcWalletState     :: !WalletState
     , wwmcConnectionsVar  :: !ConnectionsVar
+    , wwmcHashes          :: !AddrCIdHashes
     , wwmcRealModeContext :: !(RealModeContext WalletSscType)
     }
 
+newtype AddrCIdHashes = AddrCIdHashes { unAddrCIdHashes :: (IORef (Map CHash Address)) }
+
 makeLensesWith postfixLFields ''WalletWebModeContext
 
+instance HasLens AddrCIdHashes WalletWebModeContext AddrCIdHashes where
+    lensOf = wwmcHashes_L
+
+convertCIdTOAddr :: (MonadWalletWebMode m) => CId Addr -> m Address
+convertCIdTOAddr i@(CId id) = do
+    hmRef <- unAddrCIdHashes <$> view (lensOf @AddrCIdHashes)
+    maddr <- atomicModifyIORef' hmRef $ \hm ->
+      case id `M.lookup` hm of
+       Just addr -> (hm, Right addr)
+       _         -> case cIdToAddress i of
+                    -- decoding can fail, but we don't cache failures
+                      Right addr -> (M.insert id addr hm, Right addr)
+                      Left  err  -> (hm,                  Left err)
+    either (throwM . DecodeError) pure maddr
+
+convertCIdTOAddrs :: (MonadWalletWebMode m, Traversable t) => t (CId Addr) -> m (t Address)
+convertCIdTOAddrs cids = do
+    hmRef <- unAddrCIdHashes <$> view (lensOf @AddrCIdHashes)
+    maddrs <- atomicModifyIORef' hmRef $ \hm ->
+      let lookups = map (\cid@(CId h) -> (h, M.lookup h hm, cIdToAddress cid)) cids
+          hm'     = Foldable.foldl' accum hm lookups
+
+          accum m (cid, Nothing, Right addr) = M.insert cid addr m
+          accum m _                          = m
+
+          result (_, Just addr, _)   = Right addr
+          result (_, Nothing, maddr) = maddr
+
+       in (hm', map result lookups)
+
+    mapM (either (throwM . DecodeError) pure) maddrs
+
```

Comment:
Lets not forget for next steps in improvements here that we want to eliminate this cache and just use `Address` in the DB itself, not `CId Addr`.